global_planner/src/library/cell.cpp

```diff
@@ -70,14 +70,23 @@ Cell Cell::getNeighborFromYaw(double yaw) const {
 }
 
 // Returns the neighbors of the Cell whose risk influences the Cell
-std::vector<Cell> Cell::getFlowNeighbors() const {
-  return std::vector<Cell>{
-      Cell(std::tuple<int, int, int>(xIndex() + 1, yIndex(), zIndex())),
-      Cell(std::tuple<int, int, int>(xIndex() - 1, yIndex(), zIndex())),
-      Cell(std::tuple<int, int, int>(xIndex(), yIndex() + 1, zIndex())),
-      Cell(std::tuple<int, int, int>(xIndex(), yIndex() - 1, zIndex())),
-      Cell(std::tuple<int, int, int>(xIndex(), yIndex(), zIndex() + 1)),
-      Cell(std::tuple<int, int, int>(xIndex(), yIndex(), zIndex() - 1))};
+std::vector<Cell> Cell::getFlowNeighbors(int radius) const {
+  std::vector<Cell> cells;
+  auto ceilDistance = [](int radius, int x) {
+    return static_cast<int>(std::ceil(
+        std::sqrt(std::pow(double(radius), 2) - std::pow(double(x), 2))));
+  };
+  for (int x = -radius; x <= radius; x++) {
```

Comment:
Much easier to read. However, I think it isn't actually getting the radius we expect, since the z radius only knows about the y and not the x.

The naive way to do it something like this, generate a cube and filter out the non-sphere parts:
```c++
auto inSphere = [radius](int x, int y, int z,) -> bool {
    return std::ceil(std::sqrt(x*x + y*y + z*z)) >= radius;
};

for (int x = -radius; x <= radius; x++) {
    for (int y = -radius; y <= radius; y++) {
        for(int z = -radius; z <= radius; z++) {
            if (inSphere(x,y,z)) {
                cells.push_back(Cell(std::tuple<int, int, int>(xIndex() + x, yIndex() + y, zIndex() + z)));
            }
        }
    }
}
```
So we should add a unit test to make sure it gives the same result as the naive version.

But basically, I think we need to modify `ceilDistance` to take 2 arguments, so it becomes something like:

```c++
auto ceilDistance = [](int radius, int x, int y) {
    return static_cast<int>(std::ceil(std::sqrt(std::pow(double(radius), 2) - std::pow(double(x), 2) - std::pow(double(y), 2))));
};
```

And then gets used like:
```c++
int y_radius = ceilDistance(radius, x, 0);
```
```c++
int z_radius = ceilDistance(radius, x, y);
```
This way z gets correctly limited.

Reply:
@jkflying Actually, I realized that I messed up. I made a simple fix by switching `radius` with `y_radius` Would this be enough? or should I try and do the cube filtering you suggested?

Reply:
This substitutes in `z_radius = sqrt(ceil(sqrt(r^2 - x^2))^2 - y^2)`, if we can ignore the `ceil` then it is `sqrt(r^2 - x^2 - y^2)` which is correct. The `ceil` makes it over-estimate a little, since we `ceil` before the sqrt/sqr cancel out. So it could be wrong by sqrt(y_radius).

If we want to be more accurate, we could do `int z_radius = ceilDistance(hypot(x_radius, y_radius), y);` and change the lambda to accept doubles.